### PR TITLE
Add Bel solution to reverse linked list

### PR DESCRIPTION
**Pull request by masak**
Translated from the Scheme implementation.

Bel is a very beautiful, "small" language, even moreso than Scheme. But I'm not 100% confident this will work. Need to try it against a working Bel implementation before merging.

Also, should probably comment the code a bit more.

### CODE REVIEW

**Comment by masak** [comment]
Since it came up (in email ✉️), I thought I might say a few words about the subtle differences between this version and the Scheme version from which it was ported.

* Can't `define-record-printer` in Bel. Boo! 👻 But, oh! I should totally do it as a regular function; the output would look better. I need to figure out what would be an "idiomatic stringification" of this value in Bel.

* At least `prs` is a nice analogue of `fprintf` when I do so.

* In Scheme it's `(define (reverse-linked-list lst) ...)` (parens around the whole thing, like later usage); in Bel it's `(def reverse-linked-list (lst) ...)` (parens around the parameter list). I believe the reason for Bel's choice is to do with the pattern-matching abilities on the parameter list; the operator name does not participate in that.

* Bel has a `loop` macro! I use it with gusto. The syntax is `loop (var init update test . body)`, and so L6 is to be read as "initialize `node` to `lst!first` (the `first` field of `lst`); update it after each iteration to `next-node`; use as the criterion for continuing the function `idfn` (the identity function; that is, any `node` value besides `nil` means we continue)". In Scheme only the `init` logic was on the same line as the `let loop` (L30). The rest was spread out over L31 (`test`) and L35 (`update`). I think I like Bel's way better. (To be fair to Scheme, I could've defined a `loop` macro there also. But that didn't feel commensurate to the itch it was scratching, especially since the recursive helper syntax is there and is idiomatic Scheme.)

* No need for `set-cdr!` in Bel. I just do `set (cdr node)`. You can `set` a lot of things in Bel. Later we also `set lst!first`. This might seem like a small thing, but it's super-elegant and very useful/addictive. (Common Lisp has a similar construction called `setf`. [See here](https://github.com/masak/alma/issues/214#issuecomment-284537469).)

**Comment by masak** [comment]
Someone should totally make an implementation of Bel, so that I can test this code. :wink:

Actually, I'm aware of one implementation: [believe](https://github.com/luksamuk/believe) by [Lucas Vieira](https://github.com/luksamuk). It's in early stages, but it's literally the most compliant out there (that I know of). I should take it for a spin.

**Comment by masak** [comment]
> I should take it for a spin.

I did, got partway, and [submitted an issue](https://github.com/luksamuk/believe/issues/1) on his repo. A way to make friends! ❤️

**Comment by masak** [comment]
After a few more PRs to that repo, I made [the "Acknowledgements" section](https://github.com/luksamuk/believe#acknowledgements)! 🎸

**Comment by masak** [comment]
Today I rewrote the solution with a recursive help function. This was not completely necessary, but on reflection it feels closer to the "idiomatic style" of `bel.bel` itself.

This brings us further from the loop in both the TypeScript and the Scheme solution. But the point isn't to be close to them, the point is to be a "clean" expression in whatever language. An in Bel, recursion is the bee's knees.

I also had the idea to _test_ the recursive helper function in my Bel implementation. It works!

```
> (set lst '(1 2 3 4 5))
(1 2 3 4 5)
> lst
(1 2 3 4 5)
> (def rev-linked-node (node (o prev)) (if node (let next (cdr node) (set (cdr node) prev) (rev-linked-node next node)) prev))
(lit clo nil (node (o prev)) (if node (let next (cdr node) (set (cdr node) prev) (rev-linked-node next node)) prev))
> (rev-linked-node lst)
(5 4 3 2 1)
> lst
(1)
```

But I don't want to merge this PR until the whole file works.

**Comment by masak** [comment]
I was just able to run the code on (a modified version of) my Bel implementation. It output this:

```
<linked-list (5 4 3 2 1)>
```

It currently takes 6.1 seconds to run on my laptop. I believe it can run a lot faster.

**Comment by masak** [comment]
Ok, now it works on Bel's `master` branch, so let's take it out of "draft" at least.

I still would like it to be more well-commented before we merge.

File: ch2.2.3-ex7/reverse-linked-list.bel [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+(def prs-linked-list (lst)
+  (prs "<linked-list " lst!contents ">"))
+
+(def rev-linked-list (lst)
+  (let last (rev-linked-node lst!contents)
+    (set lst!contents last)))
+
+(def rev-linked-node (node (o prev))
+  (if node
```
Comment (masak) [inline ch2.2.3-ex7/reverse-linked-list.bel:11]: As much as `if node` makes the code one line shorter, I think it's more idiomatic to reverse the test and write it as `if (no node)`, with `nil` as the consequent.

The way it is not, `nil` is still the consequent, but that's implicit in the semantics of `if`, not explicit in the code. Hiding the base case like that may score points for brevity, but not for clarity. Besides, I count 43 hits on the search `if (no` in bel.bel, but I can't recall seeing the corresponding `if` form anywhere.

**Comment by masak** [comment]
I think this code is ready for merge. [It's been given a test in the Bel repository](https://github.com/masak/bel/blob/5bd8449919152c64196548eea4b06b14d42368f5/t/reverse-linked-list.bel) (an old version, I now see, before the 'no next` thing was fixed). Perhaps comments should still be lifted over to match the TypeScript version.
